src/Talis/Persona/Client/ComplexKeyCacheStorage.php

```diff
@@ -0,0 +1,58 @@
+<?php
+/**
+ * Cache HTTP requests by defining the cache key with the request headers
+ */
+namespace Talis\Persona\Client;
+use Guzzle\Plugin\Cache\DefaultCacheStorage;
+use Guzzle\Http\Message\RequestInterface;
+
+/**
+ * Define a DefaultCacheStorage implementation that also uses the
+ * request HTTP headers as the cache key
+ */
+class ComplexKeyCacheStorage extends DefaultCacheStorage
+{
+    /**
+     * Construct a instance
+     * @param CacheProvider $cache      Cache backend to use
+     * @param string        $keyPrefix  String prefix to append to the cache key
+     * @param int           $defaultTtl Integer time for the cache to live
+     */
+    public function __construct($cache, $keyPrefix = '', $defaultTtl = 3600)
+    {
+        parent::__construct($cache, $keyPrefix, $defaultTtl);
+    }
+
+    /**
+     * Hash a request URL into a string that returns cache metadata
+     *
+     * @param RequestInterface $request request to cache
+     *
+     * @return string
+     */
+    protected function getCacheKey(RequestInterface $request)
+    {
+        // Allow cache.key_filter to trim down the URL cache key by
+        // removing generate query string values (e.g. auth)
+        if ($filter = $request->getParams()->get('cache.key_filter')) {
```

Comment:
The code originates from the extended method: https://github.com/guzzle/guzzle3/blob/master/src/Guzzle/Plugin/Cache/DefaultCacheStorage.php#L184